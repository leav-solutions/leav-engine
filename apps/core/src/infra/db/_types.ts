// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}
