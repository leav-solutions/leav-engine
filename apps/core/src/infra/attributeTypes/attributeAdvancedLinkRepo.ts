// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, AqlQuery, GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IUtils} from 'utils/utils';
import {IRecord, IRecordSort} from '_types/record';
import {ILinkValue, IValueEdge} from '_types/value';
import {AttributeFormats, IAttribute} from '../../_types/attribute';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';
import {
    BASE_QUERY_IDENTIFIER,
    GetConditionPartFunc,
    IAttributeTypeRepo,
    IAttributeWithRepo
} from './attributeTypesRepo';

const VALUES_LINKS_COLLECTION = 'core_edge_values_links';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.db.dbUtils'?: IDbUtils;
    'core.utils'?: IUtils;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null,
    'core.utils': utils = null
}: IDeps = {}): IAttributeTypeRepo {
    function _getExtendedFilterPart(attributes: IAttribute[], linkedValue: GeneratedAqlQuery): GeneratedAqlQuery {
        return aql`${
            attributes
                .map(a => a.id)
                .slice(2)
                .reduce((acc, value, i) => {
                    acc.push(aql`TRANSLATE(${value}, ${i ? acc[acc.length - 1] : aql`${linkedValue}`})`);
                    if (i) {
                        acc.shift();
                    }
                    return acc;
                }, [])[0]
        }`;
    }

    const _buildLinkValue = (linkedRecord: IRecord, valueEdge: IValueEdge): ILinkValue => {
        return {
            id_value: valueEdge._key,
            value: linkedRecord,
            attribute: valueEdge.attribute,
            modified_at: valueEdge.modified_at,
            modified_by: valueEdge.modified_by,
            created_at: valueEdge.created_at,
            created_by: valueEdge.created_by,
            version: valueEdge.version ? dbUtils.convertValueVersionFromDb(valueEdge.version) : null,
            metadata: valueEdge.metadata
        };
    };

    return {
        async createValue({library, recordId, attribute, value, ctx}): Promise<ILinkValue> {
            const edgeCollec = dbService.db.edgeCollection(VALUES_LINKS_COLLECTION);

            // Create the link between records and add some metadata on it
            const edgeData: any = {
                _from: library + '/' + recordId,
                _to: attribute.linked_library + '/' + value.value,
                attribute: attribute.id,
                modified_at: value.modified_at,
                created_at: value.created_at,
                created_by: String(ctx.userId),
                modified_by: String(ctx.userId)
            };

            if (value.version) {
                edgeData.version = dbUtils.convertValueVersionToDb(value.version);
            }

            if (value.metadata) {
                edgeData.metadata = value.metadata;
            }

            const resEdge = await dbService.execute({
                query: aql`
                    INSERT ${edgeData}
                        IN ${edgeCollec}
                    RETURN NEW`,
                ctx
            });

            const savedEdge = resEdge.length ? resEdge[0] : {};

            return _buildLinkValue(utils.decomposeValueEdgeDestination(savedEdge._to), savedEdge);
        },
        async updateValue({library, recordId, attribute, value, ctx}): Promise<ILinkValue> {
            const edgeCollec = dbService.db.edgeCollection(VALUES_LINKS_COLLECTION);

            // Update value's metadata on records link
            const edgeData: any = {
                _from: library + '/' + recordId,
                _to: attribute.linked_library + '/' + value.value,
                attribute: attribute.id,
                modified_at: value.modified_at,
                created_by: value.created_by,
                modified_by: String(ctx.userId)
            };

            if (value.version) {
                edgeData.version = dbUtils.convertValueVersionToDb(value.version);
            }

            if (value.metadata) {
                edgeData.metadata = value.metadata;
            }

            const resEdge = await dbService.execute({
                query: aql`
                    UPDATE ${{_key: value.id_value}}
                        WITH ${edgeData}
                        IN ${edgeCollec}
                    RETURN NEW`,
                ctx
            });
            const savedEdge = resEdge.length ? resEdge[0] : {};

            return _buildLinkValue(utils.decomposeValueEdgeDestination(savedEdge._to), savedEdge);
        },
        async deleteValue({value, ctx}): Promise<ILinkValue> {
            const edgeCollec = dbService.db.edgeCollection(VALUES_LINKS_COLLECTION);

            // Create the link between records and add some metadata on it
            const edgeData = {
                _key: value.id_value
            };

            const resEdge = await dbService.execute({
                query: aql`
                    REMOVE ${edgeData} IN ${edgeCollec}
                    RETURN OLD`,
                ctx
            });
            const deletedEdge = resEdge.length ? resEdge[0] : {};

            return _buildLinkValue(utils.decomposeValueEdgeDestination(deletedEdge._to), deletedEdge);
        },
        async getValues({
            library,
            recordId,
            attribute,
            forceGetAllValues = false,
            options,
            ctx
        }): Promise<ILinkValue[]> {
            const edgeCollec = dbService.db.edgeCollection(VALUES_LINKS_COLLECTION);

            const queryParts = [
                aql`
                FOR linkedRecord, edge
                    IN 1 OUTBOUND ${library + '/' + recordId}
                    ${edgeCollec}
                    FILTER edge.attribute == ${attribute.id}
                `
            ];

            if (!forceGetAllValues && typeof options !== 'undefined' && options.version) {
                queryParts.push(aql`FILTER edge.version == ${options.version}`);
            }

            const limitOne = aql.literal(!attribute.multiple_values && !forceGetAllValues ? 'LIMIT 1' : '');
            queryParts.push(aql`
                ${limitOne}
                RETURN {linkedRecord, edge}
            `);
            const query = aql.join(queryParts);
            const res = await dbService.execute({query, ctx});

            return res.map(r => _buildLinkValue(dbUtils.cleanup(r.linkedRecord), r.edge));
        },
        async getValueById({library, recordId, attribute, valueId, ctx}): Promise<ILinkValue> {
            const edgeCollec = dbService.db.edgeCollection(VALUES_LINKS_COLLECTION);

            const query = aql`
                FOR linkedRecord, edge
                    IN 1 OUTBOUND ${library + '/' + recordId}
                    ${edgeCollec}
                    FILTER edge._key == ${valueId}
                    FILTER edge.attribute == ${attribute.id}
                    LIMIT 1
                    RETURN {linkedRecord, edge}
            `;

            const res = await dbService.execute({query, ctx});

            if (!res.length) {
                return null;
            }

            return _buildLinkValue(dbUtils.cleanup(res[0].edge.linkedRecord), res[0].edge);
        },
        sortQueryPart({attributes, order}: IRecordSort): AqlQuery {
            const collec = dbService.db.collection(VALUES_LINKS_COLLECTION);
            const linked = !attributes[1]
                ? {id: '_key', format: AttributeFormats.TEXT}
                : attributes[1].id === 'id'
                ? {...attributes[1], id: '_key'}
                : attributes[1];

            const linkedValue = aql`FIRST(
                FOR v, e IN 1 OUTBOUND r._id
                ${collec}
                FILTER e.attribute == ${attributes[0].id} RETURN v.${linked.id}
            )`;

            const query =
                linked.format !== AttributeFormats.EXTENDED
                    ? aql`SORT ${linkedValue} ${order}`
                    : aql`SORT ${_getExtendedFilterPart(attributes, linkedValue)} ${order}`;

            return query;
        },
        filterQueryPart(
            attributes: IAttributeWithRepo[],
            getConditionPart: GetConditionPartFunc,
            parentIdentifier = BASE_QUERY_IDENTIFIER
        ): AqlQuery {
            const collec = dbService.db.collection(VALUES_LINKS_COLLECTION);
            const linked = !attributes[1]
                ? {id: '_key', format: AttributeFormats.TEXT}
                : attributes[1].id === 'id'
                ? {...attributes[1], id: '_key'}
                : attributes[1];

            const linkIdentifier = parentIdentifier + 'v';
            const vIdentifier = aql.literal(linkIdentifier);
            const eIdentifier = aql.literal(parentIdentifier + 'e');
            const filterLinkedValue = attributes[1]._repo.filterQueryPart(
                [...attributes].splice(1),
                getConditionPart,
                linkIdentifier
            );
            const linkedValue = aql`FIRST(
                FOR ${vIdentifier}, ${eIdentifier} IN 1 OUTBOUND ${aql.literal(parentIdentifier)}._id
                    ${collec}
                    FILTER ${eIdentifier}.attribute == ${attributes[0].id}
                    ${filterLinkedValue}
                    RETURN ${vIdentifier}
            )`;

            const query =
                linked.format !== AttributeFormats.EXTENDED
                    ? aql`FILTER ${linkedValue} != null`
                    : aql`FILTER ${_getExtendedFilterPart(attributes, linkedValue)} != null`;

            return query;
        },
        async clearAllValues({attribute, ctx}): Promise<boolean> {
            return true;
        }
    };
}
