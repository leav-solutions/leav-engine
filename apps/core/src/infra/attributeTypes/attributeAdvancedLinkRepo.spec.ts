// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, Database} from 'arangojs';
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IUtils} from 'utils/utils';
import {IQueryInfos} from '_types/queryInfos';
import {AttributeTypes} from '../../_types/attribute';
import {IValue} from '../../_types/value';
import attributeAdvancedLinkRepo from './attributeAdvancedLinkRepo';
import {IAttributeTypeRepo} from './attributeTypesRepo';

describe('AttributeAdvancedLinkRepo', () => {
    const mockAttribute = {
        id: 'test_adv_link_attr',
        type: AttributeTypes.ADVANCED_LINK,
        linked_library: 'test_linked_lib',
        multiple_values: true
    };

    const savedEdgeData = {
        _id: 'core_edge_values_links/222435651',
        _rev: '_WSywvyC--_',
        _from: 'test_lib/12345',
        _to: 'test_linked_lib/987654',
        _key: '978654321',
        attribute: 'test_adv_link_attr',
        modified_at: 400999999,
        created_at: 400999999,
        modified_by: '0',
        created_by: '0',
        metadata: {my_attribute: 'metadata value'},
        version: {
            my_tree: {
                id: '1',
                library: 'test_lib'
            }
        }
    };

    const valueData: IValue = {
        id_value: '978654321',
        value: 987654,
        attribute: 'test_adv_link_attr',
        modified_at: 400999999,
        created_at: 400999999,
        modified_by: '0',
        created_by: '0'
    };

    const mockDbUtils: Mockify<IDbUtils> = {
        convertValueVersionToDb: jest.fn().mockReturnValue({my_tree: 'test_lib/1'}),
        convertValueVersionFromDb: jest.fn().mockReturnValue({
            my_tree: {
                id: '1',
                library: 'test_lib'
            }
        })
    };

    const mockUtils: Mockify<IUtils> = {
        decomposeValueEdgeDestination: jest.fn().mockReturnValue({library: 'test_linked_lib', id: '987654'})
    };

    const ctx: IQueryInfos = {
        userId: '0',
        queryId: 'AttributeAdvancedLinkRepoTest'
    };

    describe('createValue', () => {
        test('Should create a new advanced link value', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([savedEdgeData])
            };

            const attrRepo = attributeAdvancedLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils,
                'core.utils': mockUtils as IUtils
            });

            const createdVal = await attrRepo.createValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 987654,
                    modified_at: 400999999,
                    created_at: 400999999,
                    metadata: {my_attribute: 'metadata value'},
                    version: {
                        my_tree: {
                            id: '1',
                            library: 'test_lib'
                        }
                    }
                },
                ctx
            });

            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/INSERT/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(createdVal).toMatchObject({
                id_value: '978654321',
                value: {
                    library: 'test_linked_lib',
                    id: '987654'
                },
                attribute: 'test_adv_link_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0',
                metadata: {my_attribute: 'metadata value'},
                version: {
                    my_tree: {
                        id: '1',
                        library: 'test_lib'
                    }
                }
            });
        });
    });

    describe('updateValue', () => {
        test('Should update a advanced link value', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([savedEdgeData])
            };

            const attrRepo = attributeAdvancedLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils,
                'core.utils': mockUtils as IUtils
            });

            const savedVal = await attrRepo.updateValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    id_value: '987654',
                    value: 987654,
                    modified_at: 400999999,
                    metadata: {my_attribute: 'metadata value'},
                    version: {
                        my_tree: {
                            id: '1',
                            library: 'test_lib'
                        }
                    }
                },
                ctx
            });

            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/UPDATE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(savedVal).toMatchObject({
                ...valueData,
                value: {
                    library: 'test_linked_lib',
                    id: '987654'
                },
                metadata: {my_attribute: 'metadata value'},
                version: {
                    my_tree: {
                        id: '1',
                        library: 'test_lib'
                    }
                }
            });
        });
    });

    describe('deleteValue', () => {
        test('Should delete a value', async function () {
            const deletedEdgeData = {
                _id: 'core_edge_values_links/222435651',
                _rev: '_WSywvyC--_',
                _from: 'test_lib/12345',
                _to: 'test_linked_lib/987654',
                _key: '445566'
            };

            const mockDbServ: Mockify<IDbService> = {
                db: new Database(),
                execute: global.__mockPromise([deletedEdgeData])
            };

            const attrRepo = attributeAdvancedLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.utils': mockUtils as IUtils
            });

            const deletedVal = await attrRepo.deleteValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    id_value: '445566',
                    value: 987654,
                    modified_at: 400999999,
                    created_at: 400999999
                },
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/REMOVE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(deletedVal).toMatchObject({
                id_value: '445566',
                value: {
                    id: '987654',
                    library: 'test_linked_lib'
                }
            });
        });
    });

    describe('getValueByID', () => {
        test('Should return value', async function () {
            const traversalRes = [
                {
                    linkedRecord: {
                        _key: '123456',
                        _id: 'images/123456',
                        _rev: '_WgJhrXO--_',
                        created_at: 88888,
                        modified_at: 88888
                    },
                    edge: {
                        _key: '112233',
                        _id: 'core_edge_values_links/112233',
                        _from: 'ubs/222536283',
                        _to: 'images/123456',
                        _rev: '_WgJilsW--_',
                        attribute: 'test_adv_link_attr',
                        modified_at: 99999,
                        created_at: 99999,
                        modified_by: '0',
                        created_by: '0',
                        metadata: {my_attribute: 'metadata value'}
                    }
                }
            ];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalRes)
            };

            const mockCleanupRes = jest.fn().mockReturnValueOnce({
                id: 123456,
                created_at: 88888,
                modified_at: 88888
            });

            const mockDbUtilsWithCleanup = {
                ...mockDbUtils,
                cleanup: mockCleanupRes
            };

            const attrRepo = attributeAdvancedLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtilsWithCleanup as IDbUtils
            });

            const value = await attrRepo.getValueById({
                library: 'test_lib',
                recordId: '987654',
                attribute: mockAttribute,
                valueId: '112233',
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
            expect(value).toMatchObject({
                id_value: '112233',
                value: {
                    id: 123456,
                    created_at: 88888,
                    modified_at: 88888
                },
                modified_at: 99999,
                created_at: 99999,
                modified_by: '0',
                created_by: '0',
                attribute: 'test_adv_link_attr',
                metadata: {my_attribute: 'metadata value'}
            });
        });

        test("Should return null if value doesn't exists", async function () {
            const traversalRes = [];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalRes)
            };

            const attrRepo = attributeAdvancedLinkRepo({'core.infra.db.dbService': mockDbServ});

            const value = await attrRepo.getValueById({
                library: 'test_lib',
                recordId: '987654',
                attribute: mockAttribute,
                valueId: '112233',
                ctx
            });

            expect(value).toBeNull();
        });
    });
    describe('getValues', () => {
        const traversalRes = [
            {
                linkedRecord: {
                    _key: '123456',
                    _id: 'images/123456',
                    _rev: '_WgJhrXO--_',
                    created_at: 88888,
                    modified_at: 88888
                },
                edge: {
                    _key: '112233',
                    _id: 'core_edge_values_links/112233',
                    _from: 'ubs/222536283',
                    _to: 'images/123456',
                    _rev: '_WgJilsW--_',
                    attribute: 'test_adv_link_attr',
                    modified_at: 99999,
                    created_at: 99999,
                    modified_by: '0',
                    created_by: '0',
                    metadata: {my_attribute: 'metadata value'}
                }
            },
            {
                linkedRecord: {
                    _key: '123457',
                    _id: 'images/123457',
                    _rev: '_WgJhrXO--_',
                    created_at: 77777,
                    modified_at: 77777
                },
                edge: {
                    _key: '112234',
                    _id: 'core_edge_values_links/112234',
                    _from: 'ubs/222536283',
                    _to: 'images/123457',
                    _rev: '_WgJilsW--_',
                    attribute: 'test_adv_link_attr',
                    modified_at: 66666,
                    created_at: 66666,
                    modified_by: '0',
                    created_by: '0',
                    metadata: {my_attribute: 'metadata value'}
                }
            }
        ];

        test('Should return values for advanced link attribute', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalRes)
            };

            const mockCleanupRes = jest
                .fn()
                .mockReturnValueOnce({
                    id: 123456,
                    created_at: 88888,
                    modified_at: 88888
                })
                .mockReturnValueOnce({
                    id: 123457,
                    created_at: 77777,
                    modified_at: 77777
                });

            const mockDbUtilsWithCleanup = {
                ...mockDbUtils,
                cleanup: mockCleanupRes
            };

            const attrRepo = attributeAdvancedLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtilsWithCleanup as IDbUtils
            });

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttribute,
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(values.length).toBe(2);
            expect(values[0]).toMatchObject({
                id_value: '112233',
                value: {
                    id: 123456,
                    created_at: 88888,
                    modified_at: 88888
                },
                attribute: 'test_adv_link_attr',
                modified_at: 99999,
                created_at: 99999,
                modified_by: '0',
                created_by: '0',
                metadata: {my_attribute: 'metadata value'}
            });

            expect(values[1]).toMatchObject({
                id_value: '112234',
                value: {
                    id: 123457,
                    created_at: 77777,
                    modified_at: 77777
                },
                attribute: 'test_adv_link_attr',
                modified_at: 66666,
                created_at: 66666,
                modified_by: '0',
                created_by: '0'
            });
        });

        test('Should return values filtered by version', async function () {
            const traversalResWithVers = [
                {
                    linkedRecord: {
                        _key: '123457',
                        _id: 'images/123457',
                        _rev: '_WgJhrXO--_',
                        created_at: 77777,
                        modified_at: 77777
                    },
                    edge: {
                        _key: '112233',
                        _id: 'core_edge_values_links/112234',
                        _from: 'ubs/222536283',
                        _to: 'images/123457',
                        _rev: '_WgJilsW--_',
                        attribute: 'test_adv_link_attr',
                        modified_at: 66666,
                        created_at: 66666,
                        modified_by: '0',
                        created_by: '0',
                        version: {
                            my_tree: 'my_lib/1345'
                        }
                    }
                }
            ];
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalResWithVers)
            };

            const mockCleanupRes = jest.fn().mockReturnValue({
                id: 123456,
                created_at: 88888,
                modified_at: 88888
            });

            const mockDbUtilsWithCleanup = {
                ...mockDbUtils,
                cleanup: mockCleanupRes
            };

            const attrRepo = attributeAdvancedLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtilsWithCleanup as IDbUtils
            });

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttribute,
                forceGetAllValues: false,
                options: {
                    version: {
                        my_tree: {library: 'my_lib', id: '1345'}
                    }
                },
                ctx
            });

            expect(values.length).toBe(1);
            expect(values[0].id_value).toBe('112233');
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch('FILTER edge.version');
        });

        test('Should return only first value if not multiple attribute', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([traversalRes[0]])
            };

            const mockCleanupRes = jest.fn().mockReturnValue({
                id: 123456,
                created_at: 88888,
                modified_at: 88888
            });

            const mockDbUtilsWithCleanup = {
                ...mockDbUtils,
                cleanup: mockCleanupRes
            };

            const mockAttributeNotMultiVal = {
                ...mockAttribute,
                multiple_values: false
            };

            const attrRepo = attributeAdvancedLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtilsWithCleanup as IDbUtils
            });

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttributeNotMultiVal,
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch('LIMIT 1');
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(values.length).toBe(1);
            expect(values[0]).toMatchObject({
                id_value: '112233',
                value: {
                    id: 123456,
                    created_at: 88888,
                    modified_at: 88888
                },
                attribute: 'test_adv_link_attr',
                modified_at: 99999,
                created_at: 99999,
                modified_by: '0',
                created_by: '0'
            });
        });

        test('Should return all values if forced', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalRes)
            };

            const mockCleanupRes = jest
                .fn()
                .mockReturnValueOnce({
                    id: 123456,
                    created_at: 88888,
                    modified_at: 88888
                })
                .mockReturnValueOnce({
                    id: 123457,
                    created_at: 77777,
                    modified_at: 77777
                });

            const mockDbUtilsWithCleanup = {
                ...mockDbUtils,
                cleanup: mockCleanupRes
            };

            const attrRepo = attributeAdvancedLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtilsWithCleanup as IDbUtils
            });

            const mockAttrNotMultival = {
                ...mockAttribute,
                multiple_values: false
            };

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttrNotMultival,
                forceGetAllValues: true,
                ctx
            });

            expect(values.length).toBe(2);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
        });
    });
    describe('filterQueryPart', () => {
        test('Should return advanced link filter', () => {
            const mockDbServ = {
                db: new Database()
            };
            const attrRepo = attributeAdvancedLinkRepo({'core.infra.db.dbService': mockDbServ});
            const mockRepo: Mockify<IAttributeTypeRepo> = {
                filterQueryPart: jest.fn().mockReturnValue(null)
            };

            const filter = attrRepo.filterQueryPart(
                [
                    {id: 'label', type: AttributeTypes.ADVANCED_LINK, _repo: mockRepo as IAttributeTypeRepo},
                    {id: 'linked', type: AttributeTypes.SIMPLE, _repo: mockRepo as IAttributeTypeRepo}
                ],
                () => aql`== ${'MyLabel'}`
            );

            expect(filter.query).toMatch(/^FILTER/);
            expect(filter).toMatchSnapshot();
        });
    });
    describe('sortQueryPart', () => {
        test('Should return advanced link sort', () => {
            const mockDbServ = {
                db: new Database()
            };
            const attrRepo = attributeAdvancedLinkRepo({'core.infra.db.dbService': mockDbServ});
            const filter = attrRepo.sortQueryPart({
                attributes: [
                    {id: 'label', type: AttributeTypes.ADVANCED_LINK},
                    {id: 'linked', type: AttributeTypes.SIMPLE}
                ],
                order: 'ASC'
            });

            expect(filter.query).toMatch(/^SORT/);
            expect(filter).toMatchSnapshot();
        });
    });
});
