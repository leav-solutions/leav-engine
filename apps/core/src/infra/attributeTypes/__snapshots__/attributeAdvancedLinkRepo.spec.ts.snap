// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeAdvancedLinkRepo createValue Should create a new advanced link value 1`] = `
"
                    INSERT @value0
                        IN @@value1
                    RETURN NEW"
`;

exports[`AttributeAdvancedLinkRepo createValue Should create a new advanced link value 2`] = `
Object {
  "@value1": "core_edge_values_links",
  "value0": Object {
    "_from": "test_lib/12345",
    "_to": "test_linked_lib/987654",
    "attribute": "test_adv_link_attr",
    "created_at": 400999999,
    "created_by": "0",
    "metadata": Object {
      "my_attribute": "metadata value",
    },
    "modified_at": 400999999,
    "modified_by": "0",
    "version": Object {
      "my_tree": "1",
    },
  },
}
`;

exports[`AttributeAdvancedLinkRepo deleteValue Should delete a value 1`] = `
"
                    REMOVE @value0 IN @@value1
                    RETURN OLD"
`;

exports[`AttributeAdvancedLinkRepo deleteValue Should delete a value 2`] = `
Object {
  "@value1": "core_edge_values_links",
  "value0": Object {
    "_key": "445566",
  },
}
`;

exports[`AttributeAdvancedLinkRepo filterQueryPart Should return advanced link filter 1`] = `
Object {
  "_source": [Function],
  "bindVars": Object {
    "@value0": "core_edge_values_links",
    "value1": "label",
    "value2": null,
  },
  "query": "FILTER FIRST( 
                FOR rv, re IN 1 OUTBOUND r._id
                @@value0
                FILTER re.attribute == @value1
             @value2 RETURN rv ) != null",
}
`;

exports[`AttributeAdvancedLinkRepo getValueByID Should return value 1`] = `
"
                FOR linkedRecord, edge
                    IN 1 OUTBOUND @value0
                    @@value1
                    FILTER edge._key == @value2
                    FILTER edge.attribute == @value3
                    LIMIT 1
                    RETURN {linkedRecord, edge}
            "
`;

exports[`AttributeAdvancedLinkRepo getValueByID Should return value 2`] = `
Object {
  "@value1": "core_edge_values_links",
  "value0": "test_lib/987654",
  "value2": "112233",
  "value3": "test_adv_link_attr",
}
`;

exports[`AttributeAdvancedLinkRepo getValues Should return all values if forced 1`] = `
"
                FOR linkedRecord, edge
                    IN 1 OUTBOUND @value0
                    @@value1
                    FILTER edge.attribute == @value2
                 
                
                RETURN {linkedRecord, edge}
            "
`;

exports[`AttributeAdvancedLinkRepo getValues Should return only first value if not multiple attribute 1`] = `
"
                FOR linkedRecord, edge
                    IN 1 OUTBOUND @value0
                    @@value1
                    FILTER edge.attribute == @value2
                 
                LIMIT 1
                RETURN {linkedRecord, edge}
            "
`;

exports[`AttributeAdvancedLinkRepo getValues Should return only first value if not multiple attribute 2`] = `
Object {
  "@value1": "core_edge_values_links",
  "value0": "test_lib/123456",
  "value2": "test_adv_link_attr",
}
`;

exports[`AttributeAdvancedLinkRepo getValues Should return values filtered by version 1`] = `
"
                FOR linkedRecord, edge
                    IN 1 OUTBOUND @value0
                    @@value1
                    FILTER edge.attribute == @value2
                 FILTER edge.version == @value3 
                
                RETURN {linkedRecord, edge}
            "
`;

exports[`AttributeAdvancedLinkRepo getValues Should return values for advanced link attribute 1`] = `
"
                FOR linkedRecord, edge
                    IN 1 OUTBOUND @value0
                    @@value1
                    FILTER edge.attribute == @value2
                 
                
                RETURN {linkedRecord, edge}
            "
`;

exports[`AttributeAdvancedLinkRepo getValues Should return values for advanced link attribute 2`] = `
Object {
  "@value1": "core_edge_values_links",
  "value0": "test_lib/123456",
  "value2": "test_adv_link_attr",
}
`;

exports[`AttributeAdvancedLinkRepo sortQueryPart Should return advanced link sort 1`] = `
Object {
  "_source": [Function],
  "bindVars": Object {
    "@value0": "core_edge_values_links",
    "value1": "label",
    "value2": "linked",
    "value3": "ASC",
  },
  "query": "SORT FIRST(
                FOR v, e IN 1 OUTBOUND r._id
                @@value0
                FILTER e.attribute == @value1 RETURN v.@value2
            ) @value3",
}
`;

exports[`AttributeAdvancedLinkRepo updateValue Should update a advanced link value 1`] = `
"
                    UPDATE @value0
                        WITH @value1
                        IN @@value2
                    RETURN NEW"
`;

exports[`AttributeAdvancedLinkRepo updateValue Should update a advanced link value 2`] = `
Object {
  "@value2": "core_edge_values_links",
  "value0": Object {
    "_key": "987654",
  },
  "value1": Object {
    "_from": "test_lib/12345",
    "_to": "test_linked_lib/987654",
    "attribute": "test_adv_link_attr",
    "created_by": undefined,
    "metadata": Object {
      "my_attribute": "metadata value",
    },
    "modified_at": 400999999,
    "modified_by": "0",
    "version": Object {
      "my_tree": "1",
    },
  },
}
`;
